### PR TITLE
[React] Testing V

### PR DESCRIPTION
**Pull request by pabloccid**
## Summary
Contains puppeteer exercise for testing a generic form submit.

## Trello Card
https://trello.com/c/tS4p7Bn8

### CODE REVIEW

File: react/testing_5/src/App.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import React, { Component } from 'react';
+import { Provider } from 'react-redux'
+import Dashboard from './screens/Dashboard';
+import store from './redux/store';
+
+class App extends Component {
```
Comment (HarryPerez) [inline react/testing_5/src/App.js]: functional component

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import React from 'react'
+import { Field, reduxForm } from 'redux-form'
+
+import CustomField from '../../../../components/Field';
+import { required } from './utils';
+
+function SignupForm(props) {
+  const { handleSubmit } = props
```
Comment (HarryPerez) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.js]: semicolon `;`

File: react/testing_5/src/screens/Dashboard/components/SignupForm/styles.css [hunk 3]
```diff
@@ -0,0 +1,5 @@
+.user-list {
+  display: flex;
+  flex-direction: column;
+  align-items: center;
```
Comment (HarryPerez) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/styles.css]: order styles

File: react/testing_5/src/screens/Dashboard/index.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import React, { Component } from 'react';
+
+import logo from './logo.svg';
+import './styles.css';
+import SignupForm from './components/SignupForm';
```
Comment (HarryPerez) [inline react/testing_5/src/screens/Dashboard/index.js]: order imports

File: react/testing_5/src/components/Field/styles.css [hunk 5]
```diff
@@ -0,0 +1,6 @@
+.field-container {
+  display: flex;
+  align-items: center;
```
Comment (HarryPerez) [inline react/testing_5/src/components/Field/styles.css]: order imports

**Review by HarryPerez — CHANGES_REQUESTED** [review]
Woww!! Great job!!! this is so OP!! 👍 
giphy

File: react/testing_5/src/screens/Dashboard/components/SignupResult/index.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import React, { Fragment } from 'react';
+
+function SignupResult ({handleButtonClick}) {
```
Comment (damfinkel) [inline react/testing_5/src/screens/Dashboard/components/SignupResult/index.js]:
```suggestion
function SignupResult({ handleButtonClick }) {
```

File: react/testing_5/src/screens/Dashboard/index.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import SignupResult from './components/SignupResult';
+
+class Dashboard extends Component {
+  state = { submitted: false };
+
+  submitForm = values => this.setState({submitted: true});
```
Comment (damfinkel) [inline react/testing_5/src/screens/Dashboard/index.js]:
```suggestion
  submitForm = values => this.setState({ submitted: true });
```

File: react/testing_5/src/screens/Dashboard/index.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+  cleanForm = () => this.setState({submitted: false});
```
Comment (damfinkel) [inline react/testing_5/src/screens/Dashboard/index.js]:
```suggestion
  cleanForm = () => this.setState({ submitted: false });
```

File: react/testing_5/src/App.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import React, { Component } from 'react';
+import { Provider } from 'react-redux'
+import Dashboard from './screens/Dashboard';
```
Comment (mvbattan) [inline react/testing_5/src/App.js]: I'd split imports here

Comment (mvbattan) [reply react/testing_5/src/App.js]: Shouldn't this file be `src/app/index.js` ? 🤔

File: react/testing_5/src/App.test.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+const puppeteer = require('puppeteer');
```
Comment (mvbattan) [inline react/testing_5/src/App.test.js]: I'd use ES6 `import`. What do you think ?

File: react/testing_5/src/App.test.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+describe('H1 Text', () => {
+  it('h1 loads correctly', async () => {
+    let browser = await puppeteer.launch();
+    let page = await browser.newPage();
+
+    page.emulate({
+      viewport: {
+      width: 500,
+      height: 2400
```
Comment (mvbattan) [inline react/testing_5/src/App.test.js]: Check indentation here

File: react/testing_5/src/App.test.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      },
+      userAgent: ''
+    });
+
+    await page.goto('http://localhost:3000/');
+    await page.waitForSelector('.App-title');
```
Comment (mvbattan) [inline react/testing_5/src/App.test.js]: Shouldn't this be `.app-title` ? 🤔

File: react/testing_5/src/components/Field/index.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+import './styles.css';
+
+const Field = ({
```
Comment (mvbattan) [inline react/testing_5/src/components/Field/index.js]: I'd use `function` notation

File: react/testing_5/src/components/Field/index.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  input,
+  label,
+  type,
+  meta: { touched, error, warning },
+  syncError
+}) => (
+  <div className="field-container">
+    <label>{label}</label>
+    <div>
```
Comment (mvbattan) [inline react/testing_5/src/components/Field/index.js]: `Fragment` ?

File: react/testing_5/src/components/Field/index.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+      <input {...input} placeholder={label} type={type} />
+      {touched && error && <span className={`error error-${input.name}`}>{error}</span>}
```
Comment (mvbattan) [inline react/testing_5/src/components/Field/index.js]: These classNames aren't declared at `Field/styles.css` 🤔

File: react/testing_5/src/redux/store.js [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import { createStore, combineReducers } from 'redux'
+import { reducer as formReducer } from 'redux-form'
```
Comment (mvbattan) [inline react/testing_5/src/redux/store.js]: `reducer as form` ?

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.js (hunk 2 above)
Comment (mvbattan) [reply react/testing_5/src/screens/Dashboard/components/SignupForm/index.js]: I'd destructure at param level to keep consistency

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+  return (
+    <form onSubmit={handleSubmit} className="signup-form">
+      <Field label="First name" component={CustomField} name="firstName" type="text" validate={required} />
+      <Field label="Last name" component={CustomField} name="lastName" type="text" validate={required} />
+      <Field label="Email" component={CustomField} name="email" type="email" validate={required} />
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.js]: I'd extract these ones to a list. What do you think ?

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+      <button type="submit">Submit</button>
+    </form>
+  );
+}
+
+export default reduxForm({
+  form: 'SignUp'
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.js]: Form and field names should be within a constants file but I think this is ok

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js [hunk 19]
```diff
@@ -0,0 +1,114 @@
+const faker = require('faker');
+const puppeteer = require('puppeteer');
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js]: `import` ?

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+  beforeAll(async () => {
+    browser = await puppeteer.launch();
+    page = await browser.newPage();
+  
+    page.emulate({
+      viewport: {
+      width: 500,
+      height: 900
+      },
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js]: Check indentation here

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+      userAgent: ''
+    });
+    
+  });
+  
+  beforeEach(async () => await page.goto('http://localhost:3000/'));
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js]: As a nice to have, I think this URL should be in a constants file or even a `.env`. What do you think ?

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+  afterAll(() => browser.close());
+
+  it('Can submit signup form', async () => {
+    await page.waitForSelector('.signup-form');
+    await page.click("input[name=firstName]");
+    await page.type("input[name=firstName]", person.firstName);
+    await page.click("input[name=email]");
+    await page.type("input[name=email]", person.email);
+    await page.click("input[name=lastName]");
+    await page.type("input[name=lastName]", person.lastName);
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.test.js]: I think we can reuse `form` and `field` names if extracted

File: react/testing_5/src/screens/Dashboard/index.js [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+      <div className="App">
+        <header className="App-header">
+          <img src={logo} className="App-logo" alt="logo" />
+          <h1 className="App-title">Welcome to React</h1>
+          <h1 className="App-subtitle">This is a finger exercise</h1>
+        </header>
+        <p className="App-intro">
+          To get started, edit <code>src/App.js</code> and save to reload.
+        </p>
+        {!this.state.submitted ? <SignupForm onSubmit={this.submitForm} /> : <SignupResult handleButtonClick={this.cleanForm} />}
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/index.js]: Prefer positive evaluation first:
```
{this.state.submitted ? <SignupResult /> : <SignupForm />}
```

File: react/testing_5/src/screens/Dashboard/styles.css [hunk 24]
```diff
@@ -0,0 +1,28 @@
+.App {
```
Comment (mvbattan) [inline react/testing_5/src/screens/Dashboard/styles.css]: I'd write these in lowercase

File: react/testing_5/src/components/Field/index.js (hunk 15 above)
Comment (pabloccid) [reply react/testing_5/src/components/Field/index.js]: They are for selection purposes. If this was a real project, these errors would have styles.

File: react/testing_5/src/screens/Dashboard/components/SignupForm/index.js [hunk 25]
```diff
@@ -0,0 +1,21 @@
+import React from 'react'
+import { Field, reduxForm } from 'redux-form'
+
+import CustomField from '../../../../components/Field';
+import { required } from './utils';
```
Comment (wfolini) [inline react/testing_5/src/screens/Dashboard/components/SignupForm/index.js]: Split import here?

**Review by wfolini — COMMENTED** [review]
Good for me! I'd apply all other comments and go! What do you say if you install the eslint?

**Review by HarryPerez — APPROVED** [review]

**Review by damfinkel — APPROVED** [review]

File: react/testing_5/src/app/screens/Dashboard/components/SignupResult/index.js [hunk 26]
```diff
@@ -0,0 +1,19 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types';
+
+function SignupResult({ handleButtonClick }) {
```
Comment (mvbattan) [inline react/testing_5/src/app/screens/Dashboard/components/SignupResult/index.js]: `onClick` ? Remember that `handle` names would refer to instance methods / well known handlers

File: react/testing_5/src/app/screens/Dashboard/index.js [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+    const { submitted } = this.state;
+    return (
+      <div className="app">
+        <header className="app-header">
+          <img src={logo} className="app-logo" alt="logo" />
+          <h1 className="app-title">Welcome to React</h1>
+          <h1 className="app-subtitle">This is a finger exercise</h1>
+        </header>
+        {submitted ? (
+          <SignupResult handleButtonClick={this.cleanForm} />
```
Comment (mvbattan) [inline react/testing_5/src/app/screens/Dashboard/index.js]: `onClick={}` ?

File: react/testing_5/src/app/screens/Dashboard/index.js [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+
+  render() {
+    const { submitted } = this.state;
+    return (
+      <div className="app">
+        <header className="app-header">
+          <img src={logo} className="app-logo" alt="logo" />
+          <h1 className="app-title">Welcome to React</h1>
+          <h1 className="app-subtitle">This is a finger exercise</h1>
+        </header>
```
Comment (mvbattan) [inline react/testing_5/src/app/screens/Dashboard/index.js]: Does this represent a component conceptually ?

File: react/testing_5/src/app/screens/Dashboard/components/SignupForm/index.test.js [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+  beforeEach(async () => page.goto(TEST_BASE_URL));
+
+  afterAll(() => browser.close());
+
+  it('Can submit signup form', async () => {
+    await page.waitForSelector('.signup-form');
+    await page.click('input[name=firstName]');
+    await page.type('input[name=firstName]', person.firstName);
+    await page.click('input[name=email]');
+    await page.type('input[name=email]', person.email);
```
Comment (mvbattan) [inline react/testing_5/src/app/screens/Dashboard/components/SignupForm/index.test.js:40]: If the case, we need to specify that we should use `field` names from constants

Comment (pabloccid) [reply react/testing_5/src/app/screens/Dashboard/components/SignupForm/index.test.js]: Yes, of course. This is just an example for you, they will be deleted before merging

**Review by mvbattan — APPROVED** [review]

**Review by wfolini — APPROVED** [review]
Excellent work! Congratulations! 🎉 👏

**Pull request merged**